### PR TITLE
wiki: Groth16 Recursion, Continuations, Precompiles

### PR DESCRIPTION
**Pull request by sragss**

### CODE REVIEW

File: book/src/future/continuations.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Continuations via Chunking
+Today Jolt is a monolithic SNARK. RISC-V traces cannot be broken up, they must be proven monolithically or not at all. As a result, Jolt has a fixed maximum trace length that can be proved which is a function of the available RAM on the prover machine. Long term we'd like to solve this by implementing a streaming algorithm for sumcheck such that prover RAM usage is tunable with minimal performance loss. *TODO(sragss): Streaming prover link*
```
Comment (moodlezoup) [inline book/src/future/continuations.md:2]:
```suggestion
Today Jolt is a monolithic SNARK. RISC-V traces cannot be broken up, they must be proven monolithically or not at all. As a result, Jolt has a fixed maximum trace length that can be proved which is a function of the available RAM on the prover machine. Long term we'd like to solve this by implementing a streaming version of Jolt's prover such that prover RAM usage is tunable with minimal performance loss. *TODO(sragss): Streaming prover link*
```

not limited to sumcheck, we'd need a streaming version of the PCS too

File: book/src/future/continuations.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+
+Short term we're going to solve this via monolithic chunking. The plan: Take a trace of length $N$ split it into $M$ chunks of size $N/M$ and prove each independently. $N/M$ is a function of the max RAM available to the prover. 
+
+For the direct on-chain verifier there will be a cost linear in $M$ to verify. If we wrap this in Groth16 our cost will become constant but the Groth16 prover time will be linear in $M$. We believe this short-term trade-off is worthwhile for usability until we can implement the (more complicated) streaming algorithms. 
+
+## Specifics
+A generic config parameter will be added to the `Jolt` struct called `ContinuationConfig`. At the highest level, before calling `Jolt::prove` the trace will be split into `M` chunks. `Jolt::prove` will be called on each and return `RAM_final` which can be fed into `RAM_init` during the next iteration of `Jolt::prove`. The output verification will only be run for the final chunk.
```
Comment (moodlezoup) [inline book/src/future/continuations.md:9]:
```suggestion
A generic config parameter will be added to the `Jolt` struct called `ContinuationConfig`. At the highest level, before calling `Jolt::prove` the trace will be split into `M` chunks. `Jolt::prove` will be called on each and return `RAM_final` which can be fed into `RAM_init` during the next iteration of `Jolt::prove`. The [output zerocheck](https://jolt.a16zcrypto.com/how/read_write_memory.html#ouputs-and-panic) will only be run for the final chunk. 
```
"verification" if an overloaded term

File: book/src/future/precompiles.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# Precompiles
+Precompiles are highly optimized SNARK gadgets which can be invoked from the high-level programming language of the VM user. These gadgets are much more efficient for the prover than compiling down to the underlying ISA by exploiting the structure of the workload. In practice zkVMs use these for heavy cryptographic operations such as hash functions, signatures and other elliptic curve arithmetic.
```
Comment (moodlezoup) [inline book/src/future/precompiles.md:2]: nit:
```suggestion
Precompiles are highly optimized SNARK gadgets which can be invoked from the high-level programming language of the VM user. These gadgets can be much more efficient for the prover than compiling down to the underlying ISA by exploiting the structure of the workload. In practice zkVMs use these for heavy cryptographic operations such as hash functions, signatures and other elliptic curve arithmetic.
```

File: book/src/future/precompiles.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+By popular demand, Jolt will support these gadgets as well. The short term plan is to optimize for minimizing Jolt-core development resources rather than optimal prover speed.
+
+Precompile support plan:
+1. RV32 library for supported precompiles
```
Comment (moodlezoup) [inline book/src/future/precompiles.md:7]: What does this mean?

File: book/src/future/precompiles.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+2. Tracer picks up syscalls, sets relevant flag bits and loads memory accordingly
+3. Individual (uniform) Spartan instance for each precompile, repeated over `trace_length` steps
```
Comment (moodlezoup) [inline book/src/future/precompiles.md:9]: do we know if we're doing one Spartan instance per precompile? Or one for all the precompiles?

File: book/src/future/precompiles.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+4. Jolt config includes which precompiles are supported (there is some non-zero prover / verifier cost to including an empty precompile)
```
Comment (moodlezoup) [inline book/src/future/precompiles.md:10]:
```suggestion
4. Jolt config includes which precompiles are supported (there is some non-zero prover / verifier cost to including an unused precompile)
```

File: book/src/future/groth-16.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
+# Groth16 Recursion
+Jolt's verifier today is expensive. We estimate 1-5 million gas to verify within the EVM. Better batching techniques on opening proofs can bring this down 5-10x, but it will remain expensive. Further, the short-term continuations plan (INESERT LINK) causes a linear blowup depending on the count of monolithic trace chunks proven.
```
Comment (moodlezoup) [inline book/src/future/groth-16.md:2]: add link

File: book/src/future/groth-16.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+To solve these two issues we're aiming to add a configuration option to the Jolt prover with a post processing step, which creates a Groth16 proof of the Jolt verifier for constant proof size / cost (~280k gas on EVM) regardless of continuation chunk count or opening proof cost. This technique is industry standard.
+
+## Strategy
+The easiest way to understand the workload of the verifier circuit is to jump through the codebase starting at `vm/mod.rs Jolt::verify(...)`.  Verification can be split into 4 logical modules: [instruction lookups](https://jolt.a16zcrypto.com/how/instruction_lookups.html), [read-write memory](https://jolt.a16zcrypto.com/how/read_write_memory.html), [bytecode](https://jolt.a16zcrypto.com/how/bytecode.html), [r1cs](https://jolt.a16zcrypto.com/how/r1cs_constraints.html).
+
+Each of the modules do some combination of the following:
+- [Recursive sumcheck verification](https://jolt.a16zcrypto.com/background/sumcheck.html)
```
Comment (moodlezoup) [inline book/src/future/groth-16.md:10]: wdym by recursive? If you mean this in the context of GKR, we do have non-GKR sumchecks too
```suggestion
- [Sumcheck verification](https://jolt.a16zcrypto.com/background/sumcheck.html)
```

File: book/src/future/groth-16.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+After recursively verifying sumcheck, the verifier needs to compare the claimed evaluation of the sumcheck operand at a random point $r$: $S(r)$ to their own evaluation of the polynomial at $r$. Jolt does this with a combination of opening proofs over the constituent polynomials of $S$ and direct evaluations of the multi-linear extensions of those polynomials if they have sufficient structure. 
+
+## Specifics
+### Polynomial opening proof verification 
+HyperKZG is currently the optimal commitment scheme for recursion due to the requirement of only 2-pairing operations per opening proof. Unfortunately non-native field arithmetic will always be expensive within a circuit. 
+
+There are two options:
+- Sumcheck, MLE evaluations native, pairing operations non-native
+- Sumcheck, MLE evaluations non-native, pairing operations native
```
Comment (moodlezoup) [inline book/src/future/groth-16.md:22]:
```suggestion
- Sumcheck and MLE evaluations using native arithmetic, pairing operations using non-native arithmetic
- Sumcheck and MLE evaluations using non-native arithmetic, pairing operations using native arithmetic
```

File: book/src/future/groth-16.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+# Engineering Suggestions
+The Jolt codebase is rapidly undergoing improvements to reduce prover and verifier costs as well as simplify abstractions. As a result, it's recommended that each section above be built in modules that are convenient to rewire. Each part should be incrementally testable and adjustable. 
+
+A concrete example of changes to expect: we currently require 5-10 opening proofs per Jolt execution. Even for HyperKZG this requires 10-20 pairings which is prohibitively expensive. We are working on fixing this via better [batching](https://jolt.a16zcrypto.com/background/batched-openings.html).
+
+Suggested plan of attack:
+1. Circuit for HyperKZG verification
+2. Circuit for batched HyperKZG verification
+3. Circuit for RAM verification
+4. Circuit for lookups verification
```
Comment (moodlezoup) [inline book/src/future/groth-16.md:59]: imo this is more concrete:
```suggestion
3. Circuit for sumcheck verification
4. Circuit for instruction/subtable MLE evaluations
```

File: book/src/future/precompiles.md (hunk 4 above)
Comment (sragss) [reply book/src/future/precompiles.md:7]: Just an ergonomic library for syscalls in the guest.

File: book/src/future/precompiles.md (hunk 5 above)
Comment (sragss) [reply book/src/future/precompiles.md:9]: We're currently thinking on per precompile, but flexible depending on Justin's review

**Pull request merged**